tsd/src/tsd/authoring/importers/import_PDB.cpp

```diff
@@ -0,0 +1,419 @@
+  float3 position; // x, y, z coordinates
+  double occupancy; // Occupancy
+  double tempFactor; // Temperature factor
+  std::string element; // Element symbol
+  std::string charge; // Charge on the atom
+};
+
+using Atoms = std::vector<Atom>;
+using AtomsMap = std::map<std::string, Atoms>;
+
+/**
+ * Trims a string of whitespace characters.
+ *
+ * @param str The string to trim.
+ * @return The trimmed string.
+ */
+static std::string trim(const std::string &str)
+{
+  size_t first = str.find_first_not_of(" \t\n\r");
+  if (first == std::string::npos)
+    return "";
+  size_t last = str.find_last_not_of(" \t\n\r");
+  return str.substr(first, (last - first + 1));
+}
+
+/**
+ * Parses an atom line from a PDB file.
+ *
+ * @param line The line to parse.
+ * @param atomsMap The map of atoms to store the parsed atom in.
+ */
+void parseAtomLine(const std::string &line, AtomsMap &atomsMap)
+{
+  // PDB format specification:
+  // https://www.wwpdb.org/documentation/file-format-content/format33/sect9.html#ATOM
+
+  if (line.length() < 54) { // Minimum length for required fields
+    throw std::runtime_error("Line too short for ATOM record");
+  }
+
+  Atom atom;
+  atom.serial = std::stoi(line.substr(6, 5));
+  atom.altLoc = line[16];
+  atom.residueName = trim(line.substr(17, 3));
+  atom.chainID = line[21];
+  atom.residueSeq = std::stoi(line.substr(22, 4));
+  atom.iCode = line[26];
+  atom.position.x = std::stod(line.substr(30, 8));
+  atom.position.y = std::stod(line.substr(38, 8));
+  atom.position.z = std::stod(line.substr(46, 8));
+
+  // Optional fields
+  if (line.length() >= 60) {
+    atom.occupancy = std::stod(line.substr(54, 6));
+  }
+  if (line.length() >= 66) {
+    atom.tempFactor = std::stod(line.substr(60, 6));
+  }
+  if (line.length() >= 78) {
+    atom.element = trim(line.substr(76, 2));
+  }
+  if (line.length() >= 80) {
+    atom.charge = trim(line.substr(78, 2));
+  }
+
+  const std::string name = trim(line.substr(12, 4));
+  atomsMap[name].push_back(atom);
+}
+
+/**
+ * Reads a PDB file and generates a 3D representation.
+ *
+ * @param ctx Context in which to create the 3D representation.
+ * @param filename Path to the SWC file to read.
+ * @param location Node in the scene graph where the 3D representation should be
+ * added.
+ */
+void readPDBFile(Context &ctx, const char *filename, LayerNodeRef location)
+{
+  std::ifstream file(filename);
+  if (!file.is_open()) {
+    logError("Error opening file: %s", filename);
+    return;
+  }
+
+  AtomsMap atomsMap;
+  std::string line;
+
+  while (std::getline(file, line)) {
+    if (line.substr(0, 4) == KEY_ATOM || line.substr(0, 6) == KEY_HETATM) {
+      try {
+        parseAtomLine(line, atomsMap);
+      } catch (const std::exception &e) {
+        logError("Error parsing line: %s", line.c_str());
+        logError("Error: %s", e.what());
+        return;
+      }
+    }
+  }
+
+  const auto pdbLocation = ctx.defaultLayer()->insert_first_child(
```

Comment:
One recent thing I added was the ability for layer nodes to allow users to directly modify their children without needing the context or layer itself. Thus this could be instead written:

```
location->insert_first_child(...);
```